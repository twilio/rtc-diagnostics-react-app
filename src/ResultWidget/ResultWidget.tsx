import React from 'react';
import {
  Paper,
  TableContainer,
  Table,
  TableHead,
  TableBody,
  TableRow,
  TableCell,
  Tooltip,
  Typography,
  makeStyles,
  createStyles,
  Theme,
} from '@material-ui/core';
import InfoIcon from '@material-ui/icons/Info';

import ResultIcon from './ResultIcon/ResultIcon';
import { getBestEdge, getEdgeName } from '../utils';
import { TestWarnings, TestResults } from '../types';
import { darken, fade, lighten } from '@material-ui/core/styles/colorManipulator';
import { rows } from './rows';

const useStyles = makeStyles((theme: Theme) => {
  const getBackgroundColor = theme.palette.type === 'light' ? lighten : darken;
  return createStyles({
    table: {
      tableLayout: 'fixed',
      borderTop: `1px solid ${theme.palette.divider}`,
      '& td, & th': {
        width: '260px',
        borderRight: `1px solid
      ${
        // Same implementation as material-ui's table borderBottom.
        theme.palette.type === 'light'
          ? lighten(fade(theme.palette.divider, 1), 0.88)
          : darken(fade(theme.palette.divider, 1), 0.68)
      }`,
      },
      '& td:last-child, & th:last-child': {
        borderRight: 'none',
      },
      '& th, & td:first-child': {
        fontWeight: 'bold',
      },
    },
    tableCellContent: {
      display: 'flex',
      alignItems: 'center',
      justifyContent: 'space-between',
      wordBreak: 'break-word',
      '& svg': {
        fill: '#000',
      },
    },
    headerContent: {
      display: 'flex',
      '& p': {
        fontWeight: 'bold',
        marginLeft: '12px',
      },
    },
    bestEdgeCell: {
      background: getBackgroundColor(theme.palette.success.main, 0.9),
    },
    [TestWarnings.warn]: {
      background: getBackgroundColor(theme.palette.warning.main, 0.9),
    },
    [TestWarnings.error]: {
      background: getBackgroundColor(theme.palette.error.main, 0.9),
    },
  });
});

export default function ResultWidget(props: { results?: TestResults[] }) {
  const { results } = props;
  const classes = useStyles();

  const getTableCellClass = (isBestEdge: boolean, warning?: TestWarnings) => {
    if (warning?.includes('warn')) return classes[TestWarnings.warn];
    if (warning === TestWarnings.error) return classes[TestWarnings.error];
    if (isBestEdge) return classes.bestEdgeCell;
  };

  if (!results) return null;

  const bestEdge = getBestEdge(results);

  return (
    <TableContainer component={Paper}>
      <Table className={classes.table}>
        <TableHead>
          <TableRow>
            <TableCell></TableCell>
            {results.map((result) => {
              const isBestEdge = !!bestEdge && bestEdge.edge === result.edge;
              const className = getTableCellClass(isBestEdge);
              const edgeName = getEdgeName(result) + (isBestEdge ? ' (Recommended)' : '');
              return (
                <TableCell key={result.edge} className={className}>
                  <div className={classes.headerContent}>
                    <ResultIcon result={result} />
                    <Typography>{edgeName}</Typography>
                  </div>
                </TableCell>
              );
            })}
          </TableRow>
        </TableHead>
        <TableBody>
          {rows.map((row) => {
            const tooltipTitle = row.tooltipContent?.label;
            return (
              <TableRow key={row.label}>
                <TableCell>
                  <div className={classes.tableCellContent}>
                    {row.label}
                    {tooltipTitle && (
                      <Tooltip title={tooltipTitle} placement="top" interactive leaveDelay={250}>
                        <InfoIcon />
                      </Tooltip>
                    )}
                  </div>
                </TableCell>
                {results.map((result) => {
                  const value = row.getValue(result);
                  const warning = row.getWarning?.(result);
                  const className = getTableCellClass(!!bestEdge && bestEdge.edge === result.edge, warning);
                  const tooltipContent = warning ? row.tooltipContent?.[warning] : null;

                  return (
                    <TableCell key={result.edge} className={className}>
                      <div className={classes.tableCellContent}>
                        {value}
                        {tooltipContent && (
                          <Tooltip title={tooltipContent} placement="top" interactive leaveDelay={250}>
                            <InfoIcon />
                          </Tooltip>
                        )}
                      </div>
                    </TableCell>
                  );
                })}
              </TableRow>
            );
          })}
        </TableBody>
      </Table>
    </TableContainer>
  );
}
