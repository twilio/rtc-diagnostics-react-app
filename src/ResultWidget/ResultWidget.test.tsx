import React from 'react';
import ResultWidget from './ResultWidget';
import { shallow } from 'enzyme';
import { TableContainer, Table } from '@material-ui/core';

import toJson, { OutputMapper } from 'enzyme-to-json';

const testResults: any = [
  {
    region: 'roaming',
    results: {
      preflight: {
        callSid: 'CA12345',
        edge: 'ashburn',
        networkTiming: {
          peerConnection: {
            duration: 1618,
          },
        },
        samples: [
          {
            bytesReceived: 6296,
          },
        ],
        selectedEdge: 'roaming',
        selectedIceCandidatePair: {
          localCandidate: {
            relayProtocol: 'udp',
          },
        },
        stats: {
          jitter: {
            average: 5.4444,
            max: 9,
            min: 2,
          },
          mos: {
            average: 4.3868,
            max: 4.4067161340069765,
            min: 4.171315102552577,
          },
          rtt: {
            average: 69.667,
            max: 212,
            min: 54,
          },
        },
        testTiming: {
          start: 1593709997840,
          end: 1593710025637,
          duration: 27797,
        },
        totals: {
          packetsLostFraction: 2,
        },
        warnings: [{ name: 'high-latency' }],
        isTurnRequired: true,
        callQuality: 'excellent',
      },
      bitrate: {
        averageBitrate: 2875.4694113266423,
      },
    },
    errors: {},
  },
  {
    region: 'tokyo',
    results: {},
    errors: {
      preflight: {
        code: 31000,
      },
    },
  },
];

// This function is used to clean up the jest snapshot, making it easier to read.
const snapshotMapper: OutputMapper = (json) => {
  const match = json.type.match(/WithStyles\(ForwardRef\((\w+)\)\)/);
  if (match) {
    json.type = match[1];
  }
  return json;
};

describe('the ResultWidget component', () => {
  it('should not render when results are not present', () => {
    const wrapper = shallow(<ResultWidget />);
    expect(wrapper.find(TableContainer).exists()).toBe(false);
  });

  it('should render correctly when results are present', () => {
    const wrapper = shallow(<ResultWidget results={testResults} />);
    expect(toJson(wrapper.find(Table), { map: snapshotMapper })).toMatchSnapshot();
  });
});
