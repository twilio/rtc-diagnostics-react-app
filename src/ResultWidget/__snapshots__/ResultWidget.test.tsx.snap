// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the ResultWidget component should render correctly when results are present 1`] = `
<Table
  className="makeStyles-table-1"
>
  <TableHead>
    <TableRow>
      <TableCell />
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-headerContent-3"
        >
          <ResultIcon
            result={
              Object {
                "edge": "roaming",
                "errors": Object {},
                "results": Object {
                  "bitrate": Object {
                    "averageBitrate": 2875.4694113266423,
                  },
                  "preflight": Object {
                    "callQuality": "excellent",
                    "callSid": "CA12345",
                    "edge": "ashburn",
                    "isTurnRequired": true,
                    "networkTiming": Object {
                      "peerConnection": Object {
                        "duration": 1618,
                      },
                    },
                    "samples": Array [
                      Object {
                        "bytesReceived": 6296,
                      },
                    ],
                    "selectedEdge": "roaming",
                    "selectedIceCandidatePairStats": Object {
                      "localCandidate": Object {
                        "relayProtocol": "udp",
                      },
                    },
                    "stats": Object {
                      "jitter": Object {
                        "average": 5.4444,
                        "max": 9,
                        "min": 2,
                      },
                      "mos": Object {
                        "average": 4.3868,
                        "max": 4.4067161340069765,
                        "min": 4.171315102552577,
                      },
                      "rtt": Object {
                        "average": 69.667,
                        "max": 212,
                        "min": 54,
                      },
                    },
                    "testTiming": Object {
                      "duration": 27797,
                      "end": 1593710025637,
                      "start": 1593709997840,
                    },
                    "totals": Object {
                      "packetsLostFraction": 2,
                    },
                    "warnings": Array [
                      Object {
                        "name": "high-latency",
                      },
                    ],
                  },
                },
              }
            }
          />
          <Typography>
            Roaming - Ashburn (Recommended)
          </Typography>
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-headerContent-3"
        >
          <ResultIcon
            result={
              Object {
                "edge": "tokyo",
                "errors": Object {
                  "preflight": Object {
                    "code": 31000,
                  },
                },
                "results": Object {},
              }
            }
          />
          <Typography>
            Tokyo
          </Typography>
        </div>
      </TableCell>
    </TableRow>
  </TableHead>
  <TableBody>
    <TableRow
      key="Signalling Servers Reachable"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Signalling Servers Reachable
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                Tests connectivity to the Signalling servers. See
                 
                <Unknown
                  href="https://www.twilio.com/docs/voice/client/javascript/voice-client-js-and-mobile-sdks-network-connectivity-requirements#signalling-connectivity-requirements"
                >
                  connectivity requirements
                </Unknown>
                 
                for more information. Calls cannot be established without connectivity to the signalling servers
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          Yes
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-error-6"
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          No
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                The Signalling server could not be reached. Ensure your device has internet connectivity, DNS is configured correctly, and any firewalls allow access to Twilioâ€™s
                 
                <Unknown
                  href="https://www.twilio.com/docs/voice/client/javascript/voice-client-js-and-mobile-sdks-network-connectivity-requirements#signalling-connectivity-requirements"
                >
                  Signalling Servers
                </Unknown>
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
    </TableRow>
    <TableRow
      key="Media Servers Reachable"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Media Servers Reachable
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                Tests connectivity to the Media servers. See
                 
                <Unknown
                  href="https://www.twilio.com/docs/voice/client/javascript/voice-client-js-and-mobile-sdks-network-connectivity-requirements#voice-media-servers-connectivity-requirements"
                >
                  media connectivity requirements
                </Unknown>
                for more information. Unreachable media servers cause the call to drop prematurely and could cause one way audio type issues.
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-warn-5"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          Yes (TURN UDP)
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          Did not run
        </div>
      </TableCell>
    </TableRow>
    <TableRow
      key="Time to Media"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Time to Media
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                The time it takes for the media to flow between the caller and callee after the callee picks up. The lower the better. A high value indicates latency issues in the connection between the endpoints.
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          1618
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Jitter min/avg/max"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Jitter min/avg/max
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                Jitter is a measure of incoming audio packets time variance. A high value for jitter results in deterioration of sound quality and may introduce unacceptable latency.
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          2 / 5.44 / 9
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Latency (ms) min/avg/max"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Latency (ms) min/avg/max
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                Latency is the time a packet of data takes from sender to receiver. High latency leads to poor user experience as speakers unknowingly begin to talk over each other. Latency is usually attributed to slow or overloaded networks
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          54 / 70 / 212
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Packet Loss"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Packet Loss
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                The percentage of packets lost. High packet loss results in missing audio fragments leading to unintelligible speech. Packet loss is usually caused by overloaded routers
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          2%
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Bandwidth (kbps)"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Bandwidth (kbps)
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                The bandwidth test performs a symmetrical bandwidth test using a loopback loop through a TURN server.
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          2875
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Expected Audio Quality (MOS)"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Expected Audio Quality (MOS)
          <Tooltip
            interactive={true}
            leaveDelay={250}
            placement="top"
            title={
              <Unknown>
                Expected audio quality is calculated from jitter, latency, and packet loss measured. A measure of 3.5 and above is needed for good user experience.
              </Unknown>
            }
          >
            <InfoIcon />
          </Tooltip>
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          Excellent (4.39)
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
    <TableRow
      key="Call SID"
    >
      <TableCell>
        <div
          className="makeStyles-tableCellContent-2"
        >
          Call SID
        </div>
      </TableCell>
      <TableCell
        className="makeStyles-bestEdgeCell-4"
        key="roaming"
      >
        <div
          className="makeStyles-tableCellContent-2"
        >
          CA12345
        </div>
      </TableCell>
      <TableCell
        key="tokyo"
      >
        <div
          className="makeStyles-tableCellContent-2"
        />
      </TableCell>
    </TableRow>
  </TableBody>
</Table>
`;
